components/wso2is7.key.manager/src/main/java/org/wso2/is7/tenant/management/ISTenantSyncListener.java

```diff
@@ -0,0 +1,451 @@
+import org.wso2.carbon.apimgt.impl.kmclient.KMClientErrorDecoder;
+import org.wso2.carbon.apimgt.impl.utils.APIUtil;
+import org.wso2.carbon.stratos.common.beans.TenantInfoBean;
+import org.wso2.carbon.stratos.common.exception.StratosException;
+import org.wso2.carbon.stratos.common.listeners.TenantMgtListener;
+import org.wso2.carbon.tenant.mgt.internal.TenantMgtServiceComponent;
+import org.wso2.carbon.user.api.UserStoreException;
+import org.wso2.carbon.user.core.tenant.TenantManager;
+import org.wso2.is7.client.internal.ServiceReferenceHolder;
+import org.wso2.is7.client.model.TenantModel;
+import org.wso2.is7.client.model.TenantResponseModel;
+import org.wso2.is7.client.model.WSO2IS7TenantManagementClient;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+
+/**
+ * Listener class for Tenant management in IS
+ */
+public class ISTenantSyncListener implements TenantMgtListener {
+    private static final Log log = LogFactory.getLog(ISTenantSyncListener.class);
+    private static final String TENANT_PATH_PREFIX = "t/";
+    private static final String TENANT_MANAGEMENT_API_PATH = "api/server/v1";
+
+    private static final String MULTIVALUED = "multivalued";
+    private static final String MULTIVALUED_STRING = "multivalued_";
+    private static final APIManagerConfiguration apiManagerConfiguration;
+    private WSO2IS7TenantManagementClient wso2IS7TenantManagementClient;
+    TenantSharingConfigurationDTO tenantSharingConfiguration;
+    private String identityServerBaseUrl;
+    private boolean isTenantSyncEnabled = false;
+    private String identityServerAdminUsername;
+    private String identityServerAdminPassword;
+    boolean isAutoConfigureKeyManagerOfCurrentType = false;
+    boolean skipCreateDefaultResidentKm = false;
+
+    static {
+        apiManagerConfiguration = ServiceReferenceHolder.getInstance()
+                .getAPIManagerConfigurationService().getAPIManagerConfiguration();
+    }
+
+    public ISTenantSyncListener()  {
+        initializeTenantManagementClient();
+    }
+
+     private void initializeTenantManagementClient() {
+        tenantSharingConfiguration = apiManagerConfiguration
+                .getTenantSharingConfiguration(APIConstants.KeyManager.WSO2_IS7_KEY_MANAGER_TYPE);
+
+        skipCreateDefaultResidentKm = Boolean.parseBoolean(apiManagerConfiguration
+                .getFirstProperty(APIConstants.SKIP_CREATE_RESIDENT_KEY_MANAGER));
+
+        if (tenantSharingConfiguration != null && tenantSharingConfiguration.getProperties() != null) {
+            isTenantSyncEnabled = Boolean.parseBoolean(tenantSharingConfiguration.getProperties()
+                    .get(APIConstants.IS7TenantSharingConfigs.ENABLE_TENANT_SYNC));
+            isAutoConfigureKeyManagerOfCurrentType = Boolean.parseBoolean(tenantSharingConfiguration.getProperties()
+                    .get(APIConstants.IS7TenantSharingConfigs.AUTO_CONFIGURE_KEY_MANAGER));
```

Comment:
shall we rename it to more proper config name like ENABLE_KEY_MANAGER_AUTO_SETUP